### PR TITLE
[flink] fix the source data fetcher thread leak

### LINKED ISSUE #4092: [Bug] The Source Data Fetcher thread may not exit during job failover
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/paimon/issues) and found nothing similar.

### Paimon version

0.9

### Compute Engine

Flink

### Minimal reproduce step

image

If the task block at poll method, it will hung forever. Besides, this lead to the FileSourceSplitReader can not be close which will bring the memory leak

image

### What doesn't meet your expectations?

We should wakeup from the `poll` method during the shutdown

### Anything else?

_No response_

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by Aitozi**
### Purpose

Linked issue: close #4092

### CODE REVIEW

**Comment by Aitozi** [comment]
CC @JingsongLi

**Comment by JingsongLi** [comment]
Comment in issue.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/FileStoreSourceSplitReader.java [hunk 1]
```diff
@@ -86,11 +92,21 @@ public RecordsWithSplitIds<BulkFormat.RecordIterator<RowData>> fetch() throws IO
...
+
         checkSplitOrStartNext();
 
         // pool first, pool size is 1, the underlying implementation does not allow multiple batches
         // to be read at the same time
         FileStoreRecordIterator iterator = pool();
+        if (iterator == null) {
+            LOG.info("Skip waiting for object pool due to wakeup: {}", wakeup);
+            wakeup = false;
+            return new RecordsWithPausedSplit<>();
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/FileStoreSourceSplitReader.java:108]: Rename it to `EmptyRecordsWithSplitIds`?

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/FileStoreSourceSplitReader.java [hunk 2]
```diff
@@ -86,11 +92,21 @@ public RecordsWithSplitIds<BulkFormat.RecordIterator<RowData>> fetch() throws IO
             return new RecordsWithPausedSplit<>();
         }
 
+        if (wakeup) {
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/FileStoreSourceSplitReader.java:95]: Do we need to wakeup here?

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/FileStoreSourceSplitReader.java [hunk 3]
```diff
@@ -114,12 +130,16 @@ private boolean reachLimit() {
...
 
     private FileStoreRecordIterator pool() throws IOException {
-        try {
-            return this.pool.pollEntry();
-        } catch (InterruptedException e) {
-            Thread.currentThread().interrupt();
-            throw new IOException("Interrupted");
+        FileStoreRecordIterator iterator = null;
+        while (iterator == null && !wakeup) {
+            try {
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/FileStoreSourceSplitReader.java:135]: if (wakeup) {
   wakeup = false;
}

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/FileStoreSourceSplitReader.java [hunk 4]
```diff
@@ -65,6 +70,7 @@ public class FileStoreSourceSplitReader
     private RecordIterator<InternalRow> currentFirstBatch;
 
     private boolean paused;
+    private volatile boolean wakeup;
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/FileStoreSourceSplitReader.java:73]: An `AtomicBoolean` is better.

**Review by JingsongLi — CHANGES_REQUESTED** [review]

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/FileStoreSourceSplitReader.java (hunk 4 above)
Comment (Aitozi) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/FileStoreSourceSplitReader.java]: done

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/FileStoreSourceSplitReader.java (hunk 2 above)
Comment (Aitozi) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/FileStoreSourceSplitReader.java]: I think we need, and it's meant to reset the wakeup flag. If not, we should reset the `wakeup` at the end of this method. And short-cut at the beginning looks better. WDYT?

Comment (JingsongLi) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/FileStoreSourceSplitReader.java]: Maybe this should be in `pool`.

Comment (JingsongLi) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/FileStoreSourceSplitReader.java]: And you can change method name to `poll`....

Comment (Aitozi) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/FileStoreSourceSplitReader.java]: I mean not only `poll` should wakeup. I think the wakeup should work on the next `fetch` call not on the next `poll`. So after the `fetch` the wakeup should reset to normal.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/FileStoreSourceSplitReader.java (hunk 3 above)
Comment (Aitozi) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/FileStoreSourceSplitReader.java:132]: done

**Comment by JingsongLi** [comment]
+1

**Pull request merged**